src/QueryManager.ts

```diff
@@ -0,0 +1,222 @@
+
+  constructor (orbitStore: Store, extensions?: E) {
+    this._extensions = extensions || {} as E
+    this._store = orbitStore
+    this._ongoingQueries = {}
+
+    this.subscriptions = {}
+    this.statuses = {}
+  }
+
+  generateQueryRef (queries: Queries) {
+    const terms = this._extractTerms(queries)
+
+    let queryRef = JSON.stringify(terms)
+
+    if (this.subscriptions[queryRef]) return queryRef
+
+    this.subscriptions[queryRef] = { listeners: 0, terms }
+    return queryRef
+  }
+
+  _extractTerms (queries: Queries): Term[] {
+    return Object.keys(queries).sort().map(
+      (key) => ({ key, expression: queries[key](this._store.queryBuilder).expression as Expressions })
+    )
+  }
+
+  query (queryRef: string) {
+    return this._ongoingQueries[queryRef] ? this._ongoingQueries[queryRef].statusRef : this._query(queryRef)
+  }
+
+  _query (queryRef: string) {
+
+    const statusRef = this._generateStatusRef(queryRef)
+    this.statuses[statusRef] = { error: null, loading: false, listeners: 0 }
+
+    const terms = this.subscriptions[queryRef].terms
+
+    const queries: Promise<RecordObject>[] = terms
+      .map(({ key, expression }) =>
+        new Promise((resolve, reject) => {
+          this._store.query(expression)
+            .then(record => resolve({ [key]: record }))
+            .catch(reject)
+        })
+      )
+
+    // The statuses[statusRef] object can be deleted before this gets called if client unsubscribes
+    Promise.all(queries)
+      .then(() => {
+        if (this.statuses[statusRef]) {
+          this.statuses[statusRef].loading = false
+        }
+      })
+      .catch(error => {
+        if (this.statuses[statusRef]) {
+          this.statuses[statusRef].loading = false
+          this.statuses[statusRef].error = error
+        }
+      })
+
+    this._ongoingQueries[queryRef] = { queries, listeners: 0, statusRef }
+
+    return statusRef
+  }
+
+  queryCache (terms: Term[], { beforeQuery, onQuery, onError }: QueryCacheOptions<E> = {}): RecordObject | null {
+    let cancel = false
+    if (beforeQuery) {
+      for (let i = 0; i < terms.length; i++) {
+        const result = beforeQuery(terms[i].expression, this._extensions)
+        if (result === true) {
+          cancel = true
+          break
+        }
+      }
+    }
+
+    if (!cancel) {
+      try {
+        const res = terms.map(({ key, expression }) =>
+          ({ [key]: this._store.cache.query(expression) })
+        ).reduce((acc, result) => ({ ...acc, ...result }), {})
+
+        onQuery && onQuery(res, this._extensions)
+        return res
+      } catch (err) {
+        onError && onError(err, this._extensions)
+      }
+    }
+
+    return null
+  }
+
+  _generateStatusRef (queryRef: string) {
+    let i = 1
+    while (true) {
+      const statusRef = queryRef + `_${i}`
+      if (!this.statuses[statusRef]) return statusRef
+      else i++
+    }
+  }
+
+  async subscribeToFetch (queryRef: string, listener: () => void) {
```

Comment:
Where is this method going to be used? You should avoid async methods in components/hooks.